File: quick/quick.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <assert.h>


int comparador(const void* a, const void* b)
{
    return *(int* )a - *(int* )b;
}


int mediana_tres(int* v, int ini, int fim)
{
    int meio = (ini + fim) / 2;
    int med[] = { v[ini], v[meio], v[fim] };

    qsort(med, 3, sizeof(int), comparador);

    if (med[1] == v[ini]) return ini;
    if (med[1] == v[meio]) return meio;
    return fim;
}


void swap(int* v, int i, int j)
{
    int aux = v[i];
    v[i] = v[j];
    v[j] = aux;
}


int lomuto(int* v, int ini, int fim)
{
    int med = mediana_tres(v, ini, fim);
    swap(v, ini, med);

    int pivot = v[ini];
    int i = ini;

    for (int j = i + 1; j <= fim; j++) {
        if (v[j] <= pivot) {
            swap(v, ++i, j);
        }
    }

    swap(v, ini, i);
    return i;
}


int hoare(int* v, int ini, int fim)
{
    int med = mediana_tres(v, ini, fim);
    swap(v, ini, med);

    int pivot = v[ini];
    int i = ini;
    int j = fim;

    while (i <= j) {
        while (i <= j && v[i] <= pivot) i++;
        while (i <= j && v[j] > pivot) j--;

        if (i < j) swap(v, i, j);
    }

    swap(v, ini, j);
    return j;
}


void recursivo(int* v, int ini, int fim)
{
    if (ini < fim) {
        int idx_pivot = hoare(v, ini, fim);
        recursivo(v, ini, idx_pivot - 1);
        recursivo(v, idx_pivot + 1, fim);
    }
}


void iterativo(int* v, size_t tamanho)
{
    int pilha[tamanho]; // na teoria não cresce mais do que log2(tamanho)? não consegui usar a função por algum motivo então enfim
    int top = 0;

    pilha[top++] = 0;
    pilha[top++] = tamanho - 1;

    while (top > 0) {
        int fim = pilha[--top]; // tem que ser assim pra ele voltar pro anterior válido
        int ini = pilha[--top];

        int idx_pivot = hoare(v, ini, fim);
        if (ini < idx_pivot - 1) {
            pilha[top++] = ini;
            pilha[top++] = idx_pivot - 1;
        }

        if (idx_pivot + 1 < fim) {
            pilha[top++] = idx_pivot + 1;
            pilha[top++] = fim;
        }
    }
}


void print_lista(int* lista, size_t tamanho)
{
    printf("[");
    for (int i = 0; i < tamanho - 1; i++) {
        printf("%d, ", lista[i]);
    }
    if (tamanho > 0) printf("%d", lista[tamanho - 1]);
    printf("]\n");
}


bool ordenado(int* v, size_t tamanho)
{
    for (size_t i = 0; i < tamanho - 1; i++) {
        if (v[i] > v[i + 1]) return false;
    }
    return true;
}


int main(void)
{
    size_t tamanho = 321547;

    int poggers[tamanho];
    for (int i = 0; i < tamanho; i++) {
        poggers[i] = rand();
    }

    iterativo(poggers, tamanho);
    // recursivo(poggers, 0, tamanho - 1);
    assert(ordenado(poggers, tamanho));

    return 0;
}

File: quick/quick.py
from random import randint
from itertools import repeat
from collections import deque


def swap(v, i, j):
    v[i], v[j] = v[j], v[i]


def mediana_tres(v, ini, fim):
    meio = (ini + fim) // 2

    med = [v[ini], v[meio], v[fim]]
    med.sort()

    if (med[1] == v[ini]):
        return ini
    if (med[1] == v[meio]):
        return meio
    return fim


def lomuto(v, ini, fim):
    med = mediana_tres(v, ini, fim)
    swap(v, ini, med)

    pivot = v[ini]
    i = ini

    for j in range(ini + 1, fim + 1): # tinha errado isso, tem que ser de ini + 1, não ini
        if v[j] <= pivot:
            i += 1
            swap(v, i, j)
    
    swap(v, i, ini)
    return i


def hoare(v, ini, fim):
    med = mediana_tres(v, ini, fim)
    swap(v, ini, med)

    pivot = v[ini]
    i = ini + 1
    j = fim

    while i <= j:
        while i <= j and v[i] <= pivot:
            i += 1
        while i <= j and v[j] > pivot:
            j -= 1
        
        if i < j:
            swap(v, i, j)

    swap(v, j, ini)
    return j


def iterativo(v): # performance ta peso aqui mas funcionou, usando deque melhorou eu acho?
    fila = deque() # isso não é pilha pai

    fila.append(0)
    fila.append(len(v) - 1)

    while len(fila) > 0:
        i = fila.popleft()
        f = fila.popleft()

        i_pivot = hoare(v, i, f)

        if i < i_pivot - 1:
            fila.append(i)
            fila.append(i_pivot - 1)
        if i_pivot + 1 < f:
            fila.append(i_pivot + 1)
            fila.append(f)


def recursivo(v, ini, fim):
    if ini < fim:
        i_pivot = hoare(v, ini, fim)
        recursivo(v, ini, i_pivot - 1)
        recursivo(v, i_pivot + 1, fim)


def ordenado(v):
    for i in range(len(v) - 1):
        if v[i] > v[i+1]:
            return False
    return True


if __name__ == '__main__':
    a = [randint(0, 100000) for _ in repeat(None, 1001001)]
    # recursivo(a, 0, len(a) - 1)
    iterativo(a)
    assert ordenado(a)